### DISCUSSION TITLE
Degree bias of the starting node

### DISCUSSION THREAD

**Discussion by jaione2** [networkx/networkx#7127] [Q&A]
Good morning everyone, 
I am currently working on a heterogeneous network coming from https://doi.org/10.1038/s41467-021-21770-8. The point is that I have seen that a bias is generated depending on the number of nodes the starting node has. I will put an example for better understanding: Imagine that some of the starting nodes have only 1 related node, so the degree of the starting node is 1. When doing the random walk, the only node that is associated with the starting node will have a very high visitation frequency, because each time the walker is restarted, I must visit this node to enter the network. On the other hand, we have some starting nodes with a lot of nodes, let's say 100. Every time the walker starts from this node, the probability of these 100 nodes to be visited is the VF of the first example / 100. I understand that this is how the network works and it is the nature of the work. 

However, in this case, in which I am working with biological data coming from various databases, the amount of data that a node has reported conditions the posterior analysis. In a simple example, if we would like to know which are the most visited genes for a disease, if the disease has only 1 gene associated, this gene will always be in the TOP of the most visited genes. However, if the disease has 100, the probability of those 100 genes being in the TOP is much lower. At the end of the day, this fact generated a bias in which those diseases that have been more widely studied are underrepresented, while those diseases for which little is known, 1 gene show a very different pattern.

I was wondering if there is any method to try to avoid this bias of the degree of the initial node, so that the data of different diseases can be comparable and analysable together, removing this impact of the degree of the node, which is basically a limitation of the amount of data that we have for this node.  

Thank you very much in advance, 
Hope we can find a solution together, 

Jaione

**Comment by dschult** [comment]
What kind of network analysis are you doing? Often random walks are used as a proxy for finding the (long term) steady probability distribution that would occur via a random walk. PageRank is one example of looking for the steady distribution. If you try to compute the pagerank by using random walks, then the starting node will affect how long the walk takes before it settles down to the long run distribution. But the long run distribution will be the same. My understanding is that over the first years of google's pagerank usage, they moved from random walks to linear algebra, but then once the linear algebra got too big to handle, they moved back to random walks -- only with many fairly short walks rather than a few long walks. By exploring the network locally from enough nodes to cover most of the network, one can apparently get the statistics of long walks with less computation. I don't know how that was done. 

I'm not an expert in the analysis used by the genome network analysis world, and while we have people who work in that area here, it might be good to have a broader audience. In addition to this post, you might consider sending the question to the networkx discussion list. networkx-discuss@googlegroups.com

**Reply by jaione2** [reply]
Dear Dan, 
Thank you for your early answer! 
Answering your question, I am using the network to obtain a vector in which the visitation frequency of each of the nodes is stored, let's call this vector Diffusion Profile. This vector is used as and specific signature for specific diseases or chemical compounds. The point is that if a disease has degree=1, this only protein to which the disease is related will have a visitation frequency 5-6 orders higher than any other node in the network, that is why, this high frequency is generating some biases in posterior analyses. I hope I have answered your question. If not, just let me know so we can discuss it in more detail. 

Of course, I will send also the question to the networkx discussion list so we can have a broader audience. 
Really appreciate your answer! 
Best wishes, 

Jaione

**Reply by dschult** [reply]
In the long run (infinite number of steps) the visitation probability will be the same for all starting nodes (by the Perron Frobenius Theorem).  But the long run can be a different number of steps for different starting nodes. Perhaps you can reduce the bias by letting the walk continue longer for nodes with a low degree. (But I doubt degree is the only feature needed -- for example, number of nodes within 2 steps would also affect the number of steps needed in the random walk.)

**Reply by jaione2** [reply]
Thank you for the answer it makes sense what you are suggesting. It the any parameter provided by _networkx_ for adjusting the length of the random walk depending on the number of the starting nodes?

**Reply by dschult** [reply]
Which function/algorithm are you talking about, because we don't have a generic random walking function. Are you using the "pagerank" functions, or "current-flow", or "laplacian matrix". There are others too. Most compute the steady state directly instead of computing the random walk.
